R/compare_models.R

```diff
@@ -0,0 +1,182 @@
+  if (!(col_name %in% colnames(dat))) {
+    stop(paste0("The col_name `", col_name, "` does not exist in the data frame."))
+  }
+  group_vals <- dat %>%
+    dplyr::pull({{ col_name }})
+  group_vals_shuffled <- base::sample(group_vals)
+
+  data_shuffled <- dat %>%
+    dplyr::mutate(!!col_name := group_vals_shuffled)
+
+  return(data_shuffled)
+}
+
+
+#' Calculated a permuted p-value comparing two models
+#'
+#' @inheritParams compare_models
+#' @param group_1 name of one group to compare
+#' @param group_2 name of other group to compare
+#'
+#' @return numeric p-value comparing two models
+#' @export
+#' @author Begüm Topçuoğlu, \email{topcuoglu.begum@@gmail.com}
+#' @author Courtney R Armour, \email{armourc@@umich.edu}
+#'
+#' @examples
+#' df <- dplyr::tibble(
+#'   model = c("rf", "rf", "glmnet", "glmnet", "svmRadial", "svmRadial"),
+#'   AUC = c(.2, 0.3, 0.8, 0.9, 0.85, 0.95)
+#' )
+#' set.seed(123)
+#' permute_p_value(df, "AUC", "model", "rf", "glmnet", nperm = 100)
+permute_p_value <- function(merged_data, metric, group_name, group_1, group_2, nperm = 10000) {
+  # check that the metric and group exist in data
+  if (!(metric %in% colnames(merged_data))) {
+    stop(paste0("The metric `", metric, "` does not exist in the data."))
+  }
+  if (!(group_name %in% colnames(merged_data))) {
+    stop(paste0("The group_name `", group_name, "` does not exist in the data."))
+  }
+  # check that group_1 and group_2 exist in the data
+  if (!(group_1 %in% (merged_data %>% dplyr::pull(group_name)))) {
+    stop(paste0("group_1 `", group_1, "` does not exist in the data."))
+  }
+  if (!(group_2 %in% (merged_data %>% dplyr::pull(group_name)))) {
+    stop(paste0("group_2 `", group_2, "` does not exist in the data."))
+  }
+
+  # subset results to select metric and group columns and
+  # filter to only the two groups of interest
+  sub_data <- merged_data %>%
+    dplyr::select({{ metric }}, {{ group_name }}) %>%
+    dplyr::filter(.data[[group_name]] == {{ group_1 }} | .data[[group_name]] == {{ group_2 }})
+
+  # observed difference: quantify the absolute value of the difference
+  # in metric between the two groups
+  metric_obs <- get_difference(sub_data, {{ group_name }}, {{ metric }})
+
+  # shuffled difference: quantify the absolute value of the difference
+  # in metric between the two groups after shuffling group labels
+  rep_fn <- select_apply("replicate")
+  metric_null <- rep_fn(nperm,
+                        get_difference(shuffle_group(sub_data, group_name),
+                                       group_name,
+                                       metric
+                                       )
+                        )
+
+  # n: number of shuffled calculations
+  n <- length(metric_null)
+  # r: replications at least as extreme as observed effect
+  r <- sum(abs(metric_null) >= metric_obs)
+
+  # compute Monte Carlo p-value with correction (Davison & Hinkley, 1997)
+  p_value <- (r + 1) / (n + 1)
```

Comment:
We have a function [`calc_pvalue()`](https://github.com/SchlossLab/mikropml/blob/d6b511d39efd2fb6acae1d87ad238bce896a185b/R/utils.R#L235-L247) for calculating the p-value from permutation feature importance, which is the same statistical test. Do you think we should move this code there and call that function here?